### PR TITLE
Updated i18next types to v8.4 and added missing v2 version

### PR DESCRIPTION
**Pull request by GiedriusGrabauskas**
#17524 Added missing v2 version.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/i18next/i18next/blob/master/CHANGELOG.md
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by mxl** [comment]
@GiedriusGrabauskas Please ping me when PR will be ready for review.

**Comment by GiedriusGrabauskas** [comment]
@mxl, waiting for review.

**Comment by GiedriusGrabauskas** [comment]
Ping @mxl

File: types/i18next/index.d.ts [hunk 1]
```diff
@@ -1,157 +1,708 @@
...
-        fallbackLng?: FallbackLng;
-        ns?: string | string[];
-        keySeparator?: string;
-        nsSeparator?: string;
-        returnObjects?: boolean;
-        joinArrays?: string;
-        postProcess?: string | any[];
-        interpolation?: InterpolationOptions;
-        //add an indexer to assure that interpolation arguments can be passed
-        [x: string]: any;
```
Comment (mxl) [inline types/i18next/index.d.ts]: Why did you remove translation options?

**Review by mxl — CHANGES_REQUESTED** [review]

**Comment by GiedriusGrabauskas** [comment]
I failed with naming... Now I made some fixes.

File: types/i18next/index.d.ts [hunk 2]
```diff
@@ -1,157 +1,711 @@
...
+        saveMissingTo?: "current" | "all" | "fallback";
+
+        /**
+         * function(lng, ns, key, fallbackValue) { } used for custom missing key handling (needs saveMissing set to true!)
+         * @default false
+         */
+        missingKeyHandler?: false | ((lng: string, ns: string, key: string, fallbackValue: string) => void);
+
+        /**
+         * function(key) { // return value to display }
```
Comment (mxl) [inline types/i18next/index.d.ts]: Please look at [the PR](https://github.com/i18next/i18next-gitbook/pull/9) and update comment.

File: types/i18next/index.d.ts [hunk 3]
```diff
@@ -1,157 +1,711 @@
...
+
+        /**
+         * allows objects as valid translation result
+         * @default false
+         */
         returnObjects?: boolean;
-        returnedObjectHandler?: (key: string, value: string, options: any) => void;
+
+        /**
+         * function(key, value, options) {} gets called if object was passed in as key but returnObjects was set to false
```
Comment (mxl) [inline types/i18next/index.d.ts]: `function(key, value, options) {}` can be safely removed.

File: types/i18next/index.d.ts [hunk 4]
```diff
@@ -1,157 +1,711 @@
...
+         * see interpolation
+         * @default {...}
+         */
+        interpolation?: InterpolationOptions;
+
+        /**
+         * options for language detection - check documentation of plugin
+         * @default undefined
+         */
+        detection?: {};
```
Comment (mxl) [inline types/i18next/index.d.ts]: Change it to `any` as we do not have definitions for that plugin.

File: types/i18next/index.d.ts [hunk 5]
```diff
@@ -1,157 +1,711 @@
...
+
+        /**
+         * options for backend - check documentation of plugin
+         * @default undefined
+         */
+        backend?: {};
```
Comment (mxl) [inline types/i18next/index.d.ts]: Change it to `any` as we do not have definitions for that plugin.

File: types/i18next/index.d.ts [hunk 6]
```diff
@@ -1,157 +1,711 @@
...
+
+        /**
+         * options for cache layer - check documentation of plugin
+         * @default undefined
+         */
+        cache?: {};
```
Comment (mxl) [inline types/i18next/index.d.ts]: Change it to `any` as we do not have definitions for that plugin.

**Review by mxl — CHANGES_REQUESTED** [review]

File: types/i18next/index.d.ts [hunk 7]
```diff
@@ -1,157 +1,711 @@
...
+         * function(lng, ns, key, fallbackValue) { } used for custom missing key handling (needs saveMissing set to true!)
+         * @default false
+         */
+        missingKeyHandler?: false | ((lng: string, ns: string, key: string, fallbackValue: string) => void);
+
+        /**
+         * function(key) { // return value to display }
+         * @default noop
+         */
+        parseMissingKeyHandler?(key: string): string;
```
Comment (GiedriusGrabauskas) [inline types/i18next/index.d.ts]: Are you sure, that translation function returns not only `string` value?

File: types/i18next/index.d.ts (hunk 4 above)
Comment (GiedriusGrabauskas) [inline types/i18next/index.d.ts]: Can we decide that there is an `object` type here?

**Comment by GiedriusGrabauskas** [comment]
Wow @mxl, thanks for such a nice review. :+1: Now waiting for the next one.  😄

File: types/i18next/index.d.ts (hunk 4 above)
Comment (mxl) [inline types/i18next/index.d.ts]: Yes, I did not notice that definitions have TypeScript 2.3 version. We should use `object` type here and below for `backend` and `cache`.

File: types/i18next/index.d.ts (hunk 7 above)
Comment (mxl) [inline types/i18next/index.d.ts]: Yes, see https://github.com/i18next/i18next/blob/master/test/translator/translator.translate.combination.spec.js#L67

**Comment by mxl** [comment]
@GiedriusGrabauskas Ping me when you are finished.

**Comment by GiedriusGrabauskas** [comment]
Ping @mxl. I added some generics.

File: types/i18next-express-middleware/index.d.ts [hunk 8]
```diff
@@ -2,6 +2,7 @@
 // Project: http://i18next.com/
 // Definitions by: Cyril Schumacher <https://github.com/cyrilschumacher>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
```
Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: If i18next-express-middleware uses i18next v2 definitions then specifying TypeScript version 2.3 may be omitted.

File: types/i18next-sprintf-postprocessor/index.d.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@
 // Project: https://github.com/i18next/i18next-sprintf-postProcessor
 // Definitions by: Cyril Schumacher <https://github.com/cyrilschumacher>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
```
Comment (mxl) [inline types/i18next-sprintf-postprocessor/index.d.ts]: The above comment applies here too.

File: types/i18next/index.d.ts [hunk 10]
```diff
@@ -1,157 +1,601 @@
...
-        returnedObjectHandler?: (key: string, value: string, options: any) => void;
-        joinArrays?: string;
-        overloadTranslationOptionHandler?: (args: any[]) => TranslationOptions;
-        interpolation?: InterpolationOptions;
-        detection?: any;
-        backend?: any;
-        cache?: any;
+    type Callback = (error: any, t: TranslationFunction) => void;
+
+    type TranslationFunction<TResult = string, TValues extends object = object, TKeys extends string = string> = (key: TKeys | TKeys[], options?: TranslationOptions<TValues>) => TResult;
```
Comment (mxl) [inline types/i18next/index.d.ts]: It's a nice idea to restrict `t()` function to certain keys and options. `TResult` should be `any` by default (we discussed that in previous review).

File: types/i18next/index.d.ts [hunk 11]
```diff
@@ -1,157 +1,601 @@
...
+        /**
+         * The use function is there to load additional plugins to i18next.
+         * For available module see the plugins page and don't forget to read the documentation of the plugin.
+         */
+        use(module: any): i18n;
+
+        /**
+         * Please have a look at the translation functions like interpolation, formatting and plurals for more details on using it.
+         */
+        t<TResult = string, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TranslationOptions<TValues>): TResult;
```
Comment (mxl) [inline types/i18next/index.d.ts]: `TranslationFunction` should be used here (with default type parameters):
```
t: TranslationFunction;
```

File: types/i18next/index.d.ts [hunk 12]
```diff
@@ -1,157 +1,601 @@
...
+
+        /**
+         * Uses the same resolve functionality as the t function and returns true if a key exists.
+         */
+        exists<TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TranslationOptions<TValues>): boolean;
```
Comment (mxl) [inline types/i18next/index.d.ts]: Can be shortened to:
```
exists: TranslationFunction<boolean>;
```

**Review by mxl — CHANGES_REQUESTED** [review]

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (DovydasNavickas) [inline types/i18next-express-middleware/index.d.ts]: Why should it be omitted? Isn't it better to state a specific version?

File: types/i18next/index.d.ts (hunk 12 above)
Comment (DovydasNavickas) [inline types/i18next/index.d.ts]: No it cannot. It would lose generics.

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (GiedriusGrabauskas) [inline types/i18next-express-middleware/index.d.ts]: I added generics with default types from TS 2.3, so I need this line.

Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: It makes definitions not compatible with older (pre-2.1) TypeScript versions and developers using these definitions will be forced to upgrade to TypeScript 2.1+. Also these definitions do not use any of TypeScript 2.1+ features (https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-typescript-21-or-above).

File: types/i18next/index.d.ts (hunk 11 above)
Comment (GiedriusGrabauskas) [inline types/i18next/index.d.ts]: No, then you can't specify custom generic values to `t` method.

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (DovydasNavickas) [inline types/i18next-express-middleware/index.d.ts]: Default generics are used, so yeah, @GiedriusGrabauskas is right. This line is needed.

Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: @DovydasNavickas Where are they used? As I see i18next-express-middleware and i18next-node-fs-backend tsconfig maps `i18next` to `i18next/v2` which does not use any TypeScript 2.1+ features and does not specify TypeScript version in its header.

File: types/i18next/index.d.ts (hunk 11 above)
Comment (mxl) [inline types/i18next/index.d.ts]: To make this possible `TranslationFunction` should be defined like:
```
interface TranslationFunction {
    <TResult = any, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TValues): TResult
}
```

Comment (mxl) [inline types/i18next/index.d.ts]: Hmm, wait a little, I should think.

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (GiedriusGrabauskas) [inline types/i18next-express-middleware/index.d.ts]: [Travis tests](https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/253173362?utm_source=github_status&utm_medium=notification) failed if I remove these lines.

File: types/i18next/index.d.ts (hunk 11 above)
Comment (mxl) [inline types/i18next/index.d.ts]: So, `TranslationFunction` could be changed to:

```
interface TranslationFunction<T> {
    <TResult extends T = T, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TValues): TResult
}
```

All translation function occurrences should use `TranslationFunction<any>` type like:
```
t: TranslationFunction<any>;
```
and
```
getFixedT(lng: string | string[], ns?: string | string[]): TranslationFunction<any>;
```

And exists function should be defined as:

```
exists: TranslationFunction<boolean>;
```

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: Tests failed because you removed this line from i18next-node-fs-backend (it does not have `"i18next": ["i18next/v2"]` mappings in tsconfig). To make i18next-node-fs-backend use v2 definitions add the same lines as you added to 18next-express-middleware tsconfig.json.

Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: I tested locally and this does not fix failing. Somehow i18next definitions are used instead of i18next/v2. I will look at it later today.

File: types/i18next/index.d.ts (hunk 11 above)
Comment (GiedriusGrabauskas) [inline types/i18next/index.d.ts]: Same results, you can't specify generic values.

**Comment by mxl** [comment]
For `TranslateFunction<T>` you should be able to use as result type any
type that extends `T`. So if `t` is defined as`t: TranslateFunction<any>`
then it can be called like `t<string>(...)` and the result type will be
`string`, isn't it?

13 июля 2017 г. 4:01 PM пользователь "Giedrius Grabauskas" <
notifications@github.com> написал:

**Comment by DovydasNavickas** [comment]
No, it will have `any` as a generic type already setup there.
If it's defined as `TranslateFunction<any>`, you will get `TranslateFunction<any>`. It has to be defined with generic parameters in order to accept them.

File: types/i18next-express-middleware/index.d.ts (hunk 8 above)
Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: @GiedriusGrabauskas Though current build succeeds, definitions are still incorrect because i18next-express-middleware and i18next-sprintf-postprocessor modules have `"i18next": ["i18next/v2"]` path mapping so they have to use old i18next/v2 definitions which is compatible with TypeScript pre-2.1 and thus specifying `// TypeScript Version: 2.3` header is not required. But because i18next-* modules use `/// <reference types=".." />` syntax to reference i18next definitions the above path mapping is ignored (see last sentence in [I want to update a package to a new major version](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-update-a-package-to-a-new-major-version)). So here we have two options:
1. Remove `"i18next": ["i18next/v2"]` path mapping from i18next-* modules tsconfig.json and add `// TypeScript Version: 2.3` header to them. You may need to make changes to all these modules definitions to make them compatible with i18next 8.4 definitions.
2. Add `"i18next": ["i18next/v2"]` path mapping to all these modules. Remove added `// TypeScript Version: 2.3` header.
Also in both cases you need to  remove all `/// <reference types=".." />` or in some cases replace them with `import`.

Comment (mxl) [inline types/i18next-express-middleware/index.d.ts]: I prefer to choose option 2 now and upgrade i18next-* modules in separate PR.

File: types/i18next/index.d.ts (hunk 11 above)
Comment (mxl) [inline types/i18next/index.d.ts]: @GiedriusGrabauskas @DovydasNavickas please notice that `T` does not set result type, it just sets upper type bound.
Look at [TypeScript playground 1](https://www.typescriptlang.org/play/#src=interface%20TranslationFunction%3CT%3E%20%7B%0D%0A%20%20%20%20%3CTResult%20extends%20T%20%3D%20T%2C%20TKey%20extends%20string%20%3D%20string%3E(key%3A%20TKey)%3A%20TResult%0D%0A%7D%0D%0A%0D%0Alet%20t%3A%20TranslationFunction%3Cany%3E%20%3D%20(key)%20%3D%3E%20key%0D%0Alet%20b%3A%20TranslationFunction%3Cboolean%3E%20%3D%20(key)%20%3D%3E%20key%20%3D%3D%3D%20%22test%22%0D%0A%0D%0Aalert(t%3Cstring%2C%20%22key1%22%20%7C%20%22key2%22%3E(%22key1%22).length)%0D%0A%0D%0Aif%20(b%3Cboolean%2C%20string%3E(%22test%22))%20%7B%0D%0A%20%20%20%20alert(%22Success!%22)%0D%0A%7D%20). You can still specify type parameters when calling `t()` function (I simplified example a little).
Also current implementation returns from `getFixedT` `TranslationFunction` with default type arguments - `(key: string | string[], options?: TranslationOptions<object>) => any` and does not allow to specify type arguments when calling this function.

File: types/i18next/index.d.ts (hunk 12 above)
Comment (mxl) [inline types/i18next/index.d.ts]: @GiedriusGrabauskas It would be nice to make it `exists: TranslationFunction<boolean>`.

**Comment by GiedriusGrabauskas** [comment]
@mxl, I think all comments were resolved.

**Comment by mxl** [comment]
@GiedriusGrabauskas One more thing - `exists` has the same signature as `t` except it returns boolean, so we could make it `exists: TranslationFunction<boolean>`.

File: types/ng-i18next/index.d.ts [hunk 13]
```diff
@@ -2,6 +2,7 @@
 // Project: https://github.com/i18next/ng-i18next
 // Definitions by: Cyril Schumacher <https://github.com/cyrilschumacher>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
 
 /// <reference types="i18next" />
```
Comment (mxl) [inline types/ng-i18next/index.d.ts]: Here also TypeScript 2.3 header is not required, but you should remove `/// <reference ...`.

**Review by mxl — CHANGES_REQUESTED** [review]

**Comment by GiedriusGrabauskas** [comment]
@mxl, updated.

**Comment by GiedriusGrabauskas** [comment]
@mxl, updated.

**Comment by mxl** [comment]
@GiedriusGrabauskas Thanks you for a such great PR! Very good job! Approved 👍

**Comment by DovydasNavickas** [comment]
@mxl, change the status of PR review to "Approved" for @typescript-bot to approve it 👍

**Review by mxl — APPROVED** [review]

**Comment by mxl** [comment]
@DovydasNavickas I approved changes but I do not see any other buttons around here.

**Review by DovydasNavickas — APPROVED** [review]

**Comment by DovydasNavickas** [comment]
Yep, that's it 👍 Now @typescript-bot should see this.

/cc @RyanCavanaugh just to make sure bot merges, if that works already.
